### PR TITLE
Revert "Revert "validate settings and site config in DB layer, not elsewhere (#42039)" (#42282)"

### PR DESCRIPTION
**Pull request by kopancek**
This reverts commit 768f5349999128bddfc2fbdb6594d24ccb53586a.

This revives the PR
- #42039

## Test plan

Tested it out locally manually. It works ™️ 
Steps to repeat the test:

1. Use an incorrect site config, e.g. add `"foo": "bar"` somewhere to the site configuration json
2. Run `sg start enterprise`
3. You should see the following error in the logs:
```
[2022-10-14 17:23:2626.617629] [       frontend] fatal: failed to apply site config overrides: writing site config overrides to database: ConfStore.SiteCreateIfUpToDate: site configuration is invalid: (root): Additional property foo is not allowed
``` 
4. Fix the site config, remove the added invalid property
5. Try again to run `sg start enterprise`
6. Everything should work

### CODE REVIEW

File: internal/database/conf.go [hunk 1]
```diff
@@ -131,10 +133,21 @@ VALUES ('site', %s)
...
+		problemStruct, err = conf.Validate(conftypes.RawUnified{Site: contents})
+		problems = problemStruct.Messages()
+	} else {
+		problems, err = conf.ValidateSite(contents)
+	}
+	if err != nil {
+		return nil, errors.Errorf("failed to validate site configuration: %w", err)
+	} else if len(problems) > 0 {
+		return nil, errors.Errorf("site configuration is invalid: %s", strings.Join(problems, ","))
 	}
```
Comment (kopancek) [inline internal/database/conf.go:151]: I know this is ugly, but couldn't find a better way with Golang to achieve the same. Feel free to suggest a better / more readable way.

Comment (mrnugget) [reply internal/database/conf.go:151]: You can also create a separate error type here that makes this maybe a bit nicer:

```go
type siteConfigInvalidErr struct {
  problems []string
}

func (s siteConfigInvalidErr) Error() string {
  return fmt.Sprintf("site configuration is invalid: %s", strings.Join(problems, ","))
}
```

**Review by mrnugget — APPROVED** [review]
Could you reproduce the panic that lead to the revert commit? Is that fixed now?

**Comment by kopancek** [comment]
> 

Yes. The panic was caused by us trying to use uninitialized config store. Which was neatly hidden deep in the layers of code. I fixed it by the complicated logic that I introduced in [this commit](https://github.com/sourcegraph/sourcegraph/pull/42989/commits/d3bbbda44c11e36c9794c178417c22aba2339d38)

**Pull request merged**